### PR TITLE
Make Control Plane Ingress IP configurable

### LINKED ISSUE #2381: Control plane Ingress controller is tied to Bootstrap
**Component**: salt, ingress

**Problem**:

> The way the control-plane Ingress controller is currently deployed ain't great (using an `externalIP`).
> (see [this comment](https://github.com/scality/metalk8s/issues/2380#issuecomment-610396113))

This means the controller can only run on the Node with this `externalIP` value, in our case, the Bootstrap. (Note that the certificates are also only generated for this IP in their SANs)

**Resolution proposal**:

> This should be fixed somehow anyway (basically, your control-plane ingress running on a non-bootstrap node won't even work today). Either by having a specific `portmap` configuration, or by running them in the host network namespace and bind to the (host) control-plane IP only, somehow.
> (see [this comment](https://github.com/scality/metalk8s/issues/2380#issuecomment-610396113))

### PR DESCRIPTION
**Pull request by TeddyAndrieux**
**Component**:

'salt', 'tests'

**Context**: 

Being able to choose which IP will be used for Ingress Control Plane, so that user can manage them self HA of this IP through all Control Plane nodes.

**Summary**:

- Define Control Plane Ingress IP and Endpoint at a single place
- Male this Control Plane Ingress IP configurable in the Bootstrap Configuration
- Some docs and tests for this

---

Sees: #2381 

--- 

NOTE: It's only the first part of Control Plane Ingress HA, as we also want to add an optional MetalLB deployment as part of MetalK8s, so that if it's possible in the user environment HA of Control Plane Ingress can be managed by MetalLB directly.

### CODE REVIEW

File: salt/tests/unit/modules/files/test_metalk8s_network.yaml [hunk 1]
```diff
@@ -83,3 +83,47 @@ routes:
...
+
+get_control_plane_ingress_ep:
+  # 1. Nominal
+  - cp_ingress_ip_ret: 1.1.1.1
+    result: https://1.1.1.1:8443
+
+  # 2. Error when getting IP
+  - cp_ingress_ip_ret: "Error because of banana"
+    raises: true
+    result: "Error because of banana"
```
Comment (alexandre-allard) [inline salt/tests/unit/modules/files/test_metalk8s_network.yaml:137]: This error happens to me all the time, I'm glad it's tested ! :+1:

File: salt/_modules/metalk8s_network.py [hunk 2]
```diff
@@ -226,3 +226,31 @@ def routes():
...
+
+    if not isinstance(mine_ret, dict) or bootstrap_id not in mine_ret:
+        raise CommandExecutionError(
+            "Unable to get {} Control Plane IP: {}".format(bootstrap_id, mine_ret)
+        )
+
+    return mine_ret[bootstrap_id]
+
+
+def get_control_plane_ingress_ep():
```
Comment (gdemonet) [inline salt/_modules/metalk8s_network.py]: I'd rather use full words (`get_control_plane_ingress_endpoint`), especially since `ip` and `ep` are really close so less distinguishable.

File: CHANGELOG.md [hunk 3]
```diff
@@ -60,6 +60,12 @@
   LVMLogicalVolume volume in the UI
   (PR[#3410](https://github.com/scality/metalk8s/pull/3410))
 
+- [#2381](https://github.com/scality/metalk8s/issues/2381)) - Allow
+  to configure the external IP used for Control Plane Ingress, in
+  order for user to manage them self High Availability of this IP accross
+  all Control Plane nodes
```
Comment (alexandre-allard) [inline CHANGELOG.md]:
```suggestion
- [#2381](https://github.com/scality/metalk8s/issues/2381)) - Allow
  the configuration of the Control Plane Ingress' external IP, so that a
  user can manage himself the high availability of this IP on all the
  nodes of the control plane
```

File: docs/installation/bootstrap.rst [hunk 4]
```diff
@@ -80,6 +81,15 @@ notation for it's various subfields.
...
         which we do not recommend for non-experts.
 
+      For ``controlPlane`` entry an ``ingressIP`` can also be provided, this
+      IP is the one that will be used to connect to control plane components,
+      like MetalK8s-UI and all the monitoring stack. We suggest to use
+      a `Virtual IP <https://en.wikipedia.org/wiki/Virtual_IP_address>`_ that
+      will sit on a working master Node. The default value for this
+      Ingress IP is the control plane IP of the Bootstrap node (which means
+      that if you lose the Bootstrap node, you no longer have access to any
+      control plane conponent).
```
Comment (alexandre-allard) [inline docs/installation/bootstrap.rst]:
```suggestion
      For ``controlPlane`` entry, an ``ingressIP`` can also be provided. This
      IP is used to connect to all the control plane components, like MetalK8s-UI
      and the whole monitoring stack. We suggest using
      a `Virtual IP <https://en.wikipedia.org/wiki/Virtual_IP_address>`_ that
      will sit on a working master Node. The default value for this
      Ingress IP is the control plane IP of the Bootstrap node (which means
      that if you lose the Bootstrap node, you no longer have access to any
      control plane component).
```

File: docs/operation/changing_control_plane_ingress_ip.rst [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+   ``<bootstrap_node_name>`` with the actual Bootstrap Node name):
+
+   .. parsed-literal::
+
+      $ kubectl exec -n kube-system -c salt-master \\
+          --kubeconfig=/etc/kubernetes/admin.conf \\
+          salt-master-<bootstrap_node_name> -- salt-run \\
+          state.orchestrate \\
+          metalk8s.orchestrate.update-control-plane-ingress-ip \\
+          saltenv=metalk8s-|version|
```
Comment (alexandre-allard) [inline docs/operation/changing_control_plane_ingress_ip.rst]: Couldn't we programmatically retrieve the bootstrap node name using Salt instead of this ? It would be easier to just have to copy/paste the command.

File: docs/operation/changing_control_plane_ingress_ip.rst [hunk 6]
```diff
@@ -0,0 +1,43 @@
+Changing the Control Plane Ingress IP
+=====================================
+
+#. On the bootstrap node, update the ``ingressIP`` field from
+   ``network.controlPlane`` in the Bootstrap configuration file.
```
Comment (alexandre-allard) [inline docs/operation/changing_control_plane_ingress_ip.rst]:
```suggestion
   ``networks.controlPlane`` in the Bootstrap configuration file.
```

File: docs/operation/changing_control_plane_ingress_ip.rst (hunk 5 above)
Comment (TeddyAndrieux) [reply docs/operation/changing_control_plane_ingress_ip.rst]: Sure we could even do it with kubectl but ... to be consistent I used the same approach as other doc sections, but indeed, I agree we should make all Salt-master commands easy to copy/paste.

Something like this:
```
kubectl exec -n kube-system -c salt-master $(kubectl get pod -l "app.kubernetes.io/name=salt-master" --namespace=kube-system -o jsonpath='{.items[0].metadata.name}') -- blabla
```

Hmm, maybe I will change this one so at least we start something :smile:

**Review by alexandre-allard — DISMISSED** [review]

File: salt/metalk8s/addons/nginx-ingress-control-plane/certs/server.sls [hunk 7]
```diff
@@ -31,6 +30,7 @@ Create Control-Plane Ingress server private key:
     'nginx-ingress-control-plane.metalk8s-ingress.svc',
     'nginx-ingress-control-plane.metalk8s-ingress.svc.cluster.local',
     grains.metalk8s.control_plane_ip,
```
Comment (gdemonet) [inline salt/metalk8s/addons/nginx-ingress-control-plane/certs/server.sls]: Should we keep the local IP, since it was there only to represent the Service IP?

File: tests/post/steps/test_ui.py [hunk 8]
```diff
@@ -13,11 +9,9 @@ def test_ui(host):
 
 
 @then("we can reach the UI")
-def reach_UI(host):
-    ip = utils.get_grain(host, "metalk8s:control_plane_ip")
-
+def reach_UI(host, control_plane_ingress_ep):
     response = requests.get(
-        "https://{ip}:8443".format(ip=ip),
+        control_plane_ingress_ep,
```
Comment (gdemonet) [inline tests/post/steps/test_ui.py:14]: Realizing this should depend on the value computed by CSC, but to make the test more robust, we could simply find the right host + path from the Ingress object we're testing. Not related to this PR though, I'll open a debt ticket.

File: docs/installation/services.rst [hunk 9]
```diff
@@ -12,13 +12,14 @@ and can be used for operating, extending and upgrading a MetalK8s cluster.
...
-.. code-block:: shell
+.. code-block:: console
 
-    root@bootstrap $ salt-call grains.get metalk8s:control_plane_ip
-    local:
-        <the control plane IP>
+    root@bootstrap $ kubectl --kubeconfig=/etc/kubernetes/admin.conf \
+      get svc -n metalk8s-ingress ingress-nginx-control-plane-controller \
+      -o=jsonpath='{.spec.externalIPs[0]}{"\n"}'
+    <the ingress control plane IP>
```
Comment (gdemonet) [inline docs/installation/services.rst:22]: :cry:

File: CHANGELOG.md [hunk 10]
```diff
@@ -60,6 +60,12 @@
   LVMLogicalVolume volume in the UI
   (PR[#3410](https://github.com/scality/metalk8s/pull/3410))
 
+- [#2381](https://github.com/scality/metalk8s/issues/2381)) - Allow
+  the configuration of the Control Plane Ingress' external IP, so that a
+  user can manage himself the high availability of this IP on all the
+  nodes of the control plane
```
Comment (gdemonet) [inline CHANGELOG.md]: Maybe reword to:

> Allow configuring the Control Plane Ingress' external IP, to enable high availability with failover of this (virtual) IP between control plane nodes. This failover is not managed by MetalK8s.

File: docs/installation/bootstrap.rst [hunk 11]
```diff
@@ -47,6 +47,7 @@ Configuration
       networks:
         controlPlane:
           cidr: <CIDR-notation>
+          ingressIP: <IP-for-ingress>
```
Comment (gdemonet) [inline docs/installation/bootstrap.rst]: I would suggest we use two levels for this:

```yaml
networks:
  controlPlane:
    ingress:
      ip: <IP-for-ingress-controller-svc>
```

So we can add `ingress.host` (or `ingress.fqdn` :shrug:) later, and use it to sign certs and also in OIDC URLs (issuer URL, redirect URLs...).

Comment (gdemonet) [reply docs/installation/bootstrap.rst]: Also, maybe the field should state `ingressController`? Since it's not "one Ingress object" in particular... I don't know, maybe it's obvious given it's part of the `networks` section :shrug:

File: docs/installation/bootstrap.rst [hunk 12]
```diff
@@ -80,6 +81,15 @@ notation for it's various subfields.
...
         which we do not recommend for non-experts.
 
+      For ``controlPlane`` entry, an ``ingressIP`` can also be provided. This
+      IP is used to connect to all the control plane components, like
+      MetalK8s-UI and the whole monitoring stack. We suggest using
+      a `Virtual IP <https://en.wikipedia.org/wiki/Virtual_IP_address>`_ that
+      will sit on a working master Node. The default value for this
+      Ingress IP is the control plane IP of the Bootstrap node (which means
+      that if you lose the Bootstrap node, you no longer have access to any
+      control plane component).
```
Comment (gdemonet) [inline docs/installation/bootstrap.rst:93]: ...through the Ingress controller. But some components also listen on host ports directly, no? Especially K8s API, which is useful to access in case of Bootstrap issues.

File: docs/operation/changing_control_plane_ingress_ip.rst [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+#. Reconfigure Control Plane conponents:
+
+   .. parsed-literal::
+
+      $ kubectl exec -n kube-system -c salt-master \\
+          --kubeconfig=/etc/kubernetes/admin.conf \\
+          $(kubectl --kubeconfig=/etc/kubernetes/admin.conf get pod \\
+          -l "app.kubernetes.io/name=salt-master" \\
+          --namespace=kube-system -o jsonpath='{.items[0].metadata.name}')  \\
```
Comment (gdemonet) [inline docs/operation/changing_control_plane_ingress_ip.rst:38]: `-o name`?

**Review by gdemonet — CHANGES_REQUESTED** [review]
Few minor changes, most important IMO is the configuration format to build as future-proof as possible.

File: salt/metalk8s/addons/nginx-ingress-control-plane/certs/server.sls (hunk 7 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/addons/nginx-ingress-control-plane/certs/server.sls]: It's not needed indeed, I will remove it

File: tests/post/steps/test_ui.py (hunk 8 above)
Comment (TeddyAndrieux) [reply tests/post/steps/test_ui.py:14]: Hmm right this path is configurable from CSC :thinking:, indeed it will need some update to take this into account, but I will not do it in this PR, so yes if you can create a debt ticket for it :heart:

File: docs/installation/services.rst (hunk 9 above)
Comment (TeddyAndrieux) [reply docs/installation/services.rst:22]: It's just to check that it's the IP we want, you do not like it ?

File: docs/installation/bootstrap.rst (hunk 11 above)
Comment (TeddyAndrieux) [reply docs/installation/bootstrap.rst]: To me it's not really linked to the controller itself, it's like the IP you use to reach ("Ingress into") Control Plane stuffs.

For the extra layer, I'm not 100% convinced that we should do it now, I see the benefits, and what you suggest may make sense but it's not needed for the moment (and maybe it will never be, I don't know) so it makes the configuration file a bit more complex for ... no reason (at least for the moment) and changing the configuration file layout because we support new features later make sense at the end.

Honestly no strong opinion here, I can change it

File: docs/installation/bootstrap.rst (hunk 12 above)
Comment (TeddyAndrieux) [reply docs/installation/bootstrap.rst:93]: Right, but if you lose bootstrap you do not have access to OIDC so access to Kubernetes APIServer (should) do not work as well (if we generated kubeconfig specific to each user :wink:)
But agree, I will replace "control plane component" with `OIDC and various Control Plane UIs`

File: docs/operation/changing_control_plane_ingress_ip.rst (hunk 13 above)
Comment (TeddyAndrieux) [reply docs/operation/changing_control_plane_ingress_ip.rst:38]: Hmm right I could use it :thinking: 
It will have strange behavior if for whatever reason you have multiple salt-master running, maybe better to keep this one for the moment (so that this procedure may be compatible with future Bootstrap HA :smile: :wink: )

File: tests/post/steps/test_ui.py (hunk 8 above)
Comment (gdemonet) [reply tests/post/steps/test_ui.py:14]: Done: #3421

File: docs/installation/services.rst (hunk 9 above)
Comment (gdemonet) [reply docs/installation/services.rst:22]: Most of our commands are getting super verbose (same for running a command in the salt-master)... Not saying it's wrong :wink:, just that it'd be nice if we could wrap those calls in some `kubectl` plugin... :innocent:

File: docs/installation/bootstrap.rst (hunk 11 above)
Comment (gdemonet) [reply docs/installation/bootstrap.rst]: > it's not really linked to the controller itself, it's like the IP you use to reach ("Ingress into") Control Plane

OK, that makes sense indeed.

> it makes the configuration file a bit more complex for ... no reason (at least for the moment) and changing the configuration file layout because we support new features later make sense at the end.

IMHO, the added complexity today (need to break a word into two, add a newline character) is ridiculous compared to the amount of code + doc changes we need when changing the config file schema.

File: docs/operation/changing_control_plane_ingress_ip.rst (hunk 13 above)
Comment (gdemonet) [reply docs/operation/changing_control_plane_ingress_ip.rst:38]: Haha good point :ok_hand: We'll have to make these commands simpler anyway, so if we change how we deploy things, we don't have to update it in hundreds of places across our docs.

File: docs/installation/services.rst (hunk 9 above)
Comment (TeddyAndrieux) [reply docs/installation/services.rst:22]: Yess I agree

File: docs/operation/changing_control_plane_ingress_ip.rst [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+#. Check that the change is taken into account.
+
+   .. code-block:: console
+
+      $ salt-call metalk8s_network.get_control_plane_ingress_ip
+      local:
+          <my-new-ip>
+
+#. On the bootstrap node, reconfigure ApiServer:
```
Comment (gdemonet) [inline docs/operation/changing_control_plane_ingress_ip.rst]: (consistency)

```suggestion
#. On the Bootstrap node, reconfigure apiserver:
```

File: docs/operation/changing_control_plane_ingress_ip.rst [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+   .. parsed-literal::
+
+     $ salt-call state.sls \\
+         metalk8s.kubernetes.apiserver \\
+         saltenv=metalk8s-|version|
+
+#. Reconfigure Control Plane conponents:
```
Comment (gdemonet) [inline docs/operation/changing_control_plane_ingress_ip.rst]:
```suggestion
#. Reconfigure Control Plane components:
```

**Review by gdemonet — DISMISSED** [review]
(minor typos) :rocket:

**Review by gdemonet — APPROVED** [review]
